google_types/src/lib.rs

```
@@ -22,3 +44,243 @@ mod pb {
 }
 
 pub use pb::google::*;
+
+use pb::google::protobuf::Any;
+use prost::{
+    bytes::{Bytes, BytesMut},
+    Message,
+};
+use std::convert::{TryFrom, TryInto};
+use std::fmt::Display;
+use std::iter::FromIterator;
+
+fn encode_status(code: tonic::Code, message: String, details: Option<Any>) -> tonic::Status {
+    if let Some(details) = details {
+        let mut buffer = BytesMut::new();
+        let status = pb::google::rpc::Status {
+            code: code as i32,
+            message: message.clone(),
+            details: vec![details],
+        };
+
+        if status.encode(&mut buffer).is_ok() {
+            return tonic::Status::with_details(code, message, buffer.freeze());
+        }
```

While I realise it is unlikely, maybe instead of silently dropping the details, we `error!()` and return `UNKNOWN`?:

```rust
    error!(error=%err, "failed to serialise error response details");
    return tonic::Status::unknown(format!("failed to serialise server error: {}", err));
```

Reasoning being that if the client is using these details to differentiate between errors using the same status code, silently dropping the details may result in the client mapping the error incorrectly.